drake/core/test/matrix_compare_test.cc

```diff
@@ -0,0 +1,152 @@
+
+  EXPECT_TRUE(CompareMatrices(m1, m2, tolerance, MatrixCompareType::relative,
+                              &error_msg));
+
+  EXPECT_FALSE(CompareMatrices(m1, m3, tolerance, MatrixCompareType::absolute,
+                               &error_msg));
+
+  EXPECT_FALSE(CompareMatrices(m1, m3, tolerance, MatrixCompareType::relative,
+                               &error_msg));
+}
+
+// Tests absolute tolerance with real numbers.
+TEST(MatrixCompareTest, AbsoluteCompare) {
+  Eigen::MatrixXd m1(2, 2);
+  m1 << 0, 1, 2, 3;
+
+  Eigen::MatrixXd m2(2, 2);
+  m2 << 0, 1 - 1e-10, 2, 3;
+
+  Eigen::MatrixXd m3(2, 2);
+  m3 << 0, 1, 2 - 1e-8, 3;
+
+  Eigen::MatrixXd m4(2, 2);
+  m4 << 0, 1, 2, 3 - 1e-6;
+
+  double tolerance = 1e-8;
+  std::string error_msg;
+
+  // The difference between m1 and m2 is less than the tolerance.
+  // They should be considered equal.
+  EXPECT_TRUE(CompareMatrices(m1, m2, tolerance, MatrixCompareType::absolute,
+                              &error_msg));
+
+  // The difference between m1 and m3 is exactly equal to the tolerance.
+  // They should be considered equal.
+  EXPECT_TRUE(CompareMatrices(m1, m3, tolerance, MatrixCompareType::absolute,
+                              &error_msg));
+
+  // The difference between m1 and m4 is greater than the tolerance.
+  // They should be considered different.
+  EXPECT_FALSE(CompareMatrices(m1, m4, tolerance, MatrixCompareType::absolute,
+                               &error_msg));
+}
+
+// Tests absolute tolerance with NaN values
+TEST(MatrixCompareTest, AbsoluteNaNCompare) {
+  Eigen::MatrixXd m1(2, 2);
+  m1 << 0, 1, std::numeric_limits<double>::quiet_NaN(), 3;
+
+  Eigen::MatrixXd m2(2, 2);
+  m2 << 0, 1, std::numeric_limits<double>::quiet_NaN(), 3;
+
+  Eigen::MatrixXd m3(2, 2);
+  m3 << 0, 1 - 1e-10, std::numeric_limits<double>::quiet_NaN(), 3;
+
+  Eigen::MatrixXd m4(2, 2);
+  m4 << 0, 1, 2, 3;
+
+  double tolerance = 1e-8;
+  std::string error_msg;
+
+  // The difference between m1 and m2 is less than the tolerance.
+  // They should be considered equal.
+  EXPECT_TRUE(CompareMatrices(m1, m2, tolerance, MatrixCompareType::absolute,
+                              &error_msg));
+
+  // The difference between m1 and m3 is exactly equal to the tolerance.
+  // They should be considered equal.
+  EXPECT_TRUE(CompareMatrices(m1, m3, tolerance, MatrixCompareType::absolute,
+                              &error_msg));
+
+  // The difference between m1 and m4 is greater than the tolerance.
+  // They should be considered different.
+  EXPECT_FALSE(CompareMatrices(m1, m4, tolerance, MatrixCompareType::absolute,
+                               &error_msg))
+      << "m1 =\n"
+      << m1 << "\nm4 =\n"
+      << m4;
+}
+
+// Tests absolute tolerance with real numbers.
+TEST(MatrixCompareTest, RelativeCompare) {
+  Eigen::MatrixXd m1(2, 2);
+  m1 << 100, 100, 100, 100;
+
+  Eigen::MatrixXd m2(2, 2);
+  m2 << 100, 100 * 0.9, 100, 100;
+
+  double one_pct = 0.01;
```

Comment:
I removed the variables. I usually go by the rule that the fewer magic numbers present the better. However, in this case, since the variable is only used once as a method's parameter and the name of the parameter is "tolerance", I'll make an exception to this magic number rule.